### PR TITLE
Assign LastFirstTransactionID correctly

### PR DESCRIPTION
**Pull request by Ardagan**
**What changed?**
Assign LastFirstTxnID correctly for multiple branch case (reset workflow operation).

**Why?**
Establish expected behavior.

**How did you test it?**
unit tests

**Potential risks**
Reset workflow doesn't perform correctly

**Is hotfix candidate?**
No

### CODE REVIEW

File: common/persistence/history_manager.go [hunk 1]
```diff
@@ -446,7 +446,7 @@ func (m *executionManagerImpl) ReadHistoryBranchByBatch(
 
 	resp := &ReadHistoryBranchByBatchResponse{}
 	var err error
-	_, resp.History, resp.NextPageToken, resp.Size, err = m.readHistoryBranch(true, request)
+	_, resp.History, resp.TransactionIDs, resp.NextPageToken, resp.Size, err = m.readHistoryBranch(true, request)
```
Comment (yiminc) [inline common/persistence/history_manager.go:449]: This is getting out of hand. I think we need a struct for the results

File: service/history/nDCStateRebuilder.go [hunk 2]
```diff
@@ -180,6 +187,8 @@ func (r *nDCStateRebuilderImpl) rebuild(
 		return nil, 0, err
 	}
 
+	rebuiltMutableState.GetExecutionInfo().LastFirstEventTxnId = lastTxnId
```
Comment (yiminc) [inline service/history/nDCStateRebuilder.go:190]: we need a setter method on mutable state to set this

File: service/history/nDCStateRebuilder.go [hunk 3]
```diff
@@ -191,7 +200,7 @@ func (r *nDCStateRebuilderImpl) rebuild(
 func (r *nDCStateRebuilderImpl) initializeBuilders(
 	namespaceEntry *namespace.Namespace,
 	now time.Time,
-) (workflow.MutableState, workflow.MutableStateRebuilder) {
+) (*workflow.MutableStateImpl, workflow.MutableStateRebuilder) {
```
Comment (yiminc) [inline service/history/nDCStateRebuilder.go:203]: why this change? is caller now depends on the implementation detail of MutableStateImpl?

File: service/history/nDCStateRebuilder.go (hunk 2 above)
Comment (wxing1292) [reply service/history/nDCStateRebuilder.go:190]: the last txn ID should be managed by history builder, not here

File: common/persistence/dataInterfaces.go [hunk 4]
```diff
@@ -891,6 +891,8 @@ type (
 	ReadHistoryBranchByBatchResponse struct {
 		// History events by batch
 		History []*historypb.History
+		// TransactionID for relevant History batch
+		TransactionIDs []int64
```
Comment (wxing1292) [inline common/persistence/dataInterfaces.go:895]: how about changing 
```go
History []*historypb.History
TransactionIDs []int64
```

to something like
```go
History []History
```
where
```go
History struct {
  TransactionID int64
  PrevTransactionID int64
  Events *historypb.History
}

Comment (wxing1292) [reply common/persistence/dataInterfaces.go:895]: basically do not use 2 slices with equal length, but one slice of structs containing all data to be used

File: service/history/nDCStateRebuilder.go (hunk 2 above)
Comment (Ardagan) [reply service/history/nDCStateRebuilder.go:190]: Synced with Wenquan:
* use current approach overall for this PR for quick fix and unblocking of GetWorkflowExecutionHistoryReverse feature
* followup with change to move managing of transaction ids to history_builder. Which would effectively include:
  * Pass in event batch metadata to history_builder/rebuilder/etc when we apply/replay events
  * Move transaction id generation out of mutable stage
  * Make history_builder responsible for managing transaction ids

**Review by wxing1292 — APPROVED** [review]

File: common/persistence/history_manager.go (hunk 1 above)
Comment (Ardagan) [reply common/persistence/history_manager.go:449]: This is internal method used in two places only. I think looking at signature is faster for reading atm. My another concern would be that we create another "ReadHistoryBranch..." structure in code that will add even more ambiguity in naming than we have now.

**Review by yiminc — APPROVED** [review]

File: common/persistence/dataInterfaces.go (hunk 4 above)
Comment (Ardagan) [reply common/persistence/dataInterfaces.go:895]: I would suggest to make it a separate PR. It blows up into too big of a change.
